In [2]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns

### Question 1
What's the version of Pandas that you installed?

In [20]:
pd.__version__

'1.5.3'

Answer - version 1.5.3

In [21]:
# Read datafile
df = pd.read_csv("data/housing.csv")

### Question 2
How many columns are in the dataset?

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


Answer - 10 columns

### Question 3
Which columns in the dataset have missing values?

In [27]:
# Count missing values per columns
df.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

Answer - total_bedrooms has 207 missing values

### Question 4
How many unique values does the `ocean_proximity` column have?

In [28]:
df.ocean_proximity.nunique()

5

Answer - 5 unique values

### Question 5
What's the average value of the `median_house_value` for the houses located near the bay?

In [30]:
# Overall average value of median_house_value
df.median_house_value.agg('mean')

206855.81690891474

In [31]:
# ocean_proximity unique values
df.ocean_proximity.unique

<bound method Series.unique of 0        NEAR BAY
1        NEAR BAY
2        NEAR BAY
3        NEAR BAY
4        NEAR BAY
           ...   
20635      INLAND
20636      INLAND
20637      INLAND
20638      INLAND
20639      INLAND
Name: ocean_proximity, Length: 20640, dtype: object>

In [32]:
# Filter houses near bay and calculate average median_house_value
df[df['ocean_proximity'] == 'NEAR BAY']['median_house_value'].agg('mean')

259212.31179039303

Answer - Average value 259212.31179039303

### Question 6

1. Calculate the average of `total_bedrooms` column in the dataset.
2. Use the `fillna` method to fill the missing values in `total_bedrooms` with the mean value from the previous step.
3. Now, calculate the average of `total_bedrooms` again.
4. Has it changed?

In [35]:
# Calculate average total_bedrooms
df.total_bedrooms.agg('mean')

537.8705525375618

In [36]:
# set mean_bedrooms = average of total bedrooms
mean_bedrooms = df.total_bedrooms.agg('mean')

# fill in missing values for total_bedrooms with mean_bedrooms value
df.total_bedrooms.fillna(value=mean_bedrooms, inplace=True)

# Check for NULL values 
df.total_bedrooms.isnull().sum()


0

In [37]:
# Re-calculate average total_bedrooms
df.total_bedrooms.agg('mean')

537.8705525375617

Answer - No (to three decimal places)

### Question 7

1. Select all the options located on islands.
2. Select only columns `housing_median_age`, `total_rooms`, `total_bedrooms`.
3. Get the underlying NumPy array. Let's call it `X`.
4. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
5. Compute the inverse of `XTX`.
6. Create an array `y` with values `[950, 1300, 800, 1000, 1300]`.
7. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
8. What's the value of the last element of `w`?

In [39]:
# Filter only islands
df_islands = df[df['ocean_proximity'] == 'ISLAND']

In [42]:
# Select housing_median_age, total_rooms, total_bedrooms
df_islands = df_islands[['housing_median_age', 'total_rooms', 'total_bedrooms']]

In [43]:
df_islands.head()

housing_median_age  total_rooms  total_bedrooms
8314                27.0       1675.0           521.0
8315                52.0       2359.0           591.0
8316                52.0       2127.0           512.0
8317                52.0        996.0           264.0
8318                29.0        716.0           214.0

In [49]:
# DataFrame to Numpy array
X = df_islands.to_numpy()

In [50]:
X

array([[  27., 1675.,  521.],
       [  52., 2359.,  591.],
       [  52., 2127.,  512.],
       [  52.,  996.,  264.],
       [  29.,  716.,  214.]])

In [51]:
# Matrix multiplication of X, transpose X
XTX = np.dot(X,X.T)

In [52]:
XTX

array([[3077795., 4260640., 3830881., 1807248., 1311577.],
       [4260640., 5916866., 5322889., 2508292., 1817026.],
       [3830881., 5322889., 4788977., 2256364., 1634008.],
       [1807248., 2508292., 2256364., 1064416.,  771140.],
       [1311577., 1817026., 1634008.,  771140.,  559293.]])

In [62]:
XTX_inverse = np.linalg.inv(XTX)

In [63]:
XTX_inverse

array([[ 3.88139994e+10, -1.84879374e+10,  2.90187188e+09,
         7.54139977e+10, -1.43414843e+11],
       [-5.94615266e+10,  3.50542952e+10, -1.17064716e+10,
        -1.11656742e+11,  2.13707758e+11],
       [ 4.70979471e+10, -2.96946763e+10,  1.13531824e+10,
         8.73300933e+10, -1.67553508e+11],
       [ 5.18302291e+10, -2.08132982e+10, -3.04261339e+08,
         1.02934124e+11, -1.94961295e+11],
       [-1.06904957e+11,  4.49229529e+10, -1.52268678e+09,
        -2.11164367e+11,  4.00350590e+11]])

In [64]:
y = np.array([950, 1300, 800, 1000, 1300])

In [65]:
y

array([ 950, 1300,  800, 1000, 1300])

In [70]:
XTX_multiply = np.outer(XTX_inverse,X.T)

In [75]:
w = np.outer(XTX_multiply,y)

In [76]:
w

array([[9.95579085e+14, 1.36237138e+15, 8.38382387e+14, 1.04797798e+15,
        1.36237138e+15],
       [1.91741157e+15, 2.62382636e+15, 1.61466238e+15, 2.01832797e+15,
        2.62382636e+15],
       [1.91741157e+15, 2.62382636e+15, 1.61466238e+15, 2.01832797e+15,
        2.62382636e+15],
       ...,
       [1.94730527e+17, 2.66473353e+17, 1.63983602e+17, 2.04979502e+17,
        2.66473353e+17],
       [1.00407928e+17, 1.37400323e+17, 8.45540447e+16, 1.05692556e+17,
        1.37400323e+17],
       [8.13912750e+16, 1.11377534e+17, 6.85400211e+16, 8.56750263e+16,
        1.11377534e+17]])